src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,411 @@
+		}
+		this.start = null;
+		this.startTime = startTime;
+		this.includeStart = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param endID the message ID to end at (exclusive)
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(long endID) {
+		return endAt(channel.getMessageByID(endID));
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param msg the message ID to end at (exclusive)
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(IMessage msg) {
+		return endAt(msg, false);
+	}
+
+	/**
+	 * Specifies the time to end at.
+	 *
+	 * @param endTime the time to end at (exclusive)
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime) {
+		return endAt(endTime, false);
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param endID     the message ID to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(long endID, boolean inclusive) {
+		return endAt(channel.getMessageByID(endID), inclusive);
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param msg       the message ID to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(IMessage msg, boolean inclusive) {
+		if (msg == null || !channel.equals(msg.getChannel())) {
+			if (lenient) return this;
+			throw new IllegalArgumentException("Message is null or does not exist in channel " + channel.getStringID() + "!");
+		}
+		this.endTime = msg.getTimestamp();
+		this.includeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the time to end at.
+	 *
+	 * @param endTime   the time to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime, boolean inclusive) {
+		if (endTime == null) {
+			if (lenient) return this;
+			throw new IllegalArgumentException("End time is null!");
+		}
+		this.endTime = endTime;
+		this.includeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the maximum number of messages to get. Defaults to -1 (no limit).
+	 *
+	 * @param limit max number of messages to get
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder withMaxCount(int limit) {
+		if (limit < 0 && !lenient) throw new IllegalArgumentException("Limit is negative!");
+		this.limit = limit;
+		return this;
+	}
+
+	/**
+	 * Returns a stream of the message history.
+	 *
+	 * @return the stream of the message history
+	 * @see MessageHistoryIterator
+	 */
+	public Stream<IMessage> stream() {
+		buildRange();
+		return new MessageHistoryIterator(range).stream();
+	}
+
+	/**
+	 * Returns an iterator for this message history.
+	 * NOTE: This object can be put directly in a for-each loop without using this method.
```

Comment:
Probably artifact of my bad design making the Iterator thing Iterable. Will remove